oauth2/oauth2_auth_code_test.go

```diff
@@ -11,6 +11,7 @@ import (
 	"errors"
 	"fmt"
```

Comment:
This diff is quite messy. The two large changes I did:

- Refactored the tests so that we can use real databases in our oauth2 tests. This is required to test concurrent token refreshes which isn't possible in SQLite as SQLite can't handle concurrency.
- Added a large chunk of test cases to cover various token rotation scenarios